smartsim/launchable/baseJobGroup.py

```diff
@@ -36,23 +35,31 @@ def __iter__(self) -> t.Iterator[BaseJob]:
         """Allows iteration over the jobs in the collection."""
         return iter(self.jobs)
 
-    def __getitem__(self, idx: int) -> BaseJob:
-        """Retrieves the job at the specified index (idx)."""
-        return self.jobs[idx]
-
-    def __setitem__(self, idx: int, value: BaseJob) -> None:
+    @t.overload
+    def __setitem__(self, idx: int, value: BaseJob) -> None: ...
```

Comment:
why is setitem repeated twice and what is t.overload?

Reply:
`__setitem__` is actually defined here 3 times! The first two definitions give the different signatures to the type checker to provide more information about how a method should be used. Only the last one is the one that matters for runtime though.

Python does not have proper method overloading (each method has one, and exactly one, definition), but the [`@t.overload`](https://docs.python.org/3/library/typing.html#typing.overload) decorator informs the type checker that depending on how the method is called there are really many (in this case two) discrete signatures. This allows for python to "spoof" true method overloading at type check time.

Now if you index into the job group with an `int` mypy knows that you should be trying to assign a `BaseJob`, but if you index into the job group with a `slice`, mypy knows that it should expect an `Iterable[BaseJob]`